addons/website_version/models/experiment.py

```diff
@@ -0,0 +1,300 @@
+            'variations': [{'name': 'master', 'url': 'http://localhost/master'}]
+        }
+        version_list = vals.get('experiment_version_ids', [])
+        for version in version_list:
+            name = self.env['website_version.version'].browse([version[2]['version_id']])[0].name
+            #We must give a URL for each version in the experiment
+            exp['variations'].append({'name': name, 'url': 'http://localhost/' + name})
+        if not version_list:
+            raise Warning(_("You must select at least one version in your experiment."))
+        google_id = self.env['google.management'].create_an_experiment(exp, vals['website_id'])
+        if not google_id:
+            raise Warning(_("Please verify you give the authorizations to use google analytics api ..."))
+        vals['google_id'] = google_id
+        return super(Experiment, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        state = vals.get('state')
+        for exp in self:
+            if state and exp.state == 'ended':
+                raise Warning(_("You cannot modify an ended experiment."))
+            elif state == 'ended':
+                #temp is the data to send to Googe
+                temp = {
+                    'name': exp.name,
+                    'status': state,
+                    'variations': [{'name': 'master', 'url': 'http://localhost/master'}],
+                }
+                for exp_v in exp.experiment_version_ids:
+                    temp['variations'].append({'name': exp_v.version_id.name, 'url': 'http://localhost/'+exp_v.version_id.name})
+                #to check the constraints before to write on the google analytics account
+                self.env['google.management'].update_an_experiment(temp, exp.google_id, exp.website_id.id)
+        return super(Experiment, self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        for exp in self:
+            self.env['google.management'].delete_an_experiment(exp.google_id, exp.website_id.id)
+        return super(Experiment, self).unlink()
+
+    @api.multi
+    def update_goals(self):
+        gm_obj = self.env['google.management']
+        goals_obj = self.env['website_version.goals']
+        website_id = self.env.context.get('website_id')
+        if not website_id:
+            raise Warning("You must specify the website.")
+        for goal in gm_obj.get_goal_info(website_id)[1]['items']:
+            if not goals_obj.search([('name', '=', goal['name'])]):
+                vals = {'name': goal['name'], 'google_ref': 'ga:goal' + goal['id'] + 'Completions'}
+                goals_obj.create(vals)
+
+    @api.model
+    def check_no_overlap(self, version_ids):
+        #Check if version_ids don't overlap with running experiments and return the name of the experiment if there 's an overlap
+        version_keys = set([v['key'] for v in self.env['ir.ui.view'].search_read([('version_id', 'in', version_ids)], ['key'])])
```

Comment:
version_keys = self.env['ir.ui.view'].search([('version_id', 'in', version_ids)]).mapped('key') ?